### PR TITLE
docs: add open edx event guide

### PR DESCRIPTION
**Pull request by mariajgrimaldi**
**Description:** 
This PR adds a how-to for new event creation. Its main purpose is to serve as a guide for new contributions to the library.

**Dependencies:**
None

**Installation instructions:** 
Just documentation.

**Testing instructions:**
Just documentation.

**Reviewers:**
- [ ] tag reviewer 
- [ ] tag reviewer 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** 
What else should we add?

### CODE REVIEW

File: docs/how-to/create-new-event.rst
Comment (robrap) [inline docs/how-to/create-new-event.rst:45]: One proposal would be to add even more info about subdomains here. It could include a list of the in-use agreed upon subdomains. It might provide examples of possible other subdomains, but which would need agreement upon first use. Here is [an event design github issue](https://github.com/openedx/openedx-events/issues/72) where Dave is discussing subdomains. I'm hoping we can capture/summarize/consolidate/reduce the amount of discussion around subdomains as we move forward.

File: docs/how-to/create-new-event.rst
Comment (robrap) [inline docs/how-to/create-new-event.rst:48]: Maybe include a code example with the actual annotations, and potentially a link to a real event.

File: docs/how-to/create-new-event.rst
Comment (robrap) [inline docs/how-to/create-new-event.rst:51]: Do we have separate documentation for the implementation? Just in edx-platform?

**Review by robrap — COMMENTED** [review]
Thank you! A nice start to kick this off.

**Comment by mariajgrimaldi** [comment]
Hi there @robrap! Thanks for the feedback. I'll be addressing your comments on Monday. Sorry for the delay!

File: docs/how-to/create-new-event.rst
Comment (mariajgrimaldi) [reply docs/how-to/create-new-event.rst:51]: For now we just have a [guide](https://github.com/openedx/edx-platform/tree/master/docs/guides/hooks) in edx-platform. Should we move those guides to each repository? That might have more sense.

File: docs/how-to/create-new-event.rst
Comment (mariajgrimaldi) [reply docs/how-to/create-new-event.rst:48]: Good idea!

File: docs/how-to/create-new-event.rst
Comment (robrap) [reply docs/how-to/create-new-event.rst:45]: Sorry for the delay. I'm not sure. I guess I'd start with something small (a couple of sentences?) and only break it out if that seems necessary.

File: docs/how-to/create-new-event.rst
Comment (robrap) [reply docs/how-to/create-new-event.rst:51]: For now, maybe we just link to the edx-platform guide as an example, but not try to keep a comprehensive list here?

**Comment by mariajgrimaldi** [comment]
FYI @felipemontoya

File: docs/how-to/create-new-event.rst
Comment (mariajgrimaldi) [inline docs/how-to/create-new-event.rst:36]: According to the confluence page, there's another domain: Enterprise. But I don't know much about it & the page is still TBD in its definitions. I'll look it up later to fill this up

File: docs/how-to/create-new-event.rst
Comment (rgraber) [inline docs/how-to/create-new-event.rst:43]: Somewhere in here I think we may want to talk about the difference between state-transfer events and notification events, and how the data attributes differ between the two. I found this to be a pretty useful guide:  https://medium.com/swlh/event-notification-vs-event-carried-state-transfer-2e4fdf8f6662

Comment (mariajgrimaldi) [reply docs/how-to/create-new-event.rst:43]: You're right! Thank you for the suggestion! I'll try to pick up this work during the week and take a look at those resources

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @mariajgrimaldi!

As a core committer in this repo, you can merge this once the pull request is approved per the [core committer reviewer requirements](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Open+edX+Core+Committers#Reviews-and-Ownership) and according to the agreement with your edX Champion.


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by robrap** [comment]
@mariajgrimaldi: Someone recently thoughts some new events defined in this library would be on the event bus. I haven't quite gotten to the bottom of the confusion, but when I went to send a link to some docs, I realized that this doc never landed. I wonder how quickly we could land this, even if we'd want to improve it in the future. Thanks.

**Comment by mariajgrimaldi** [comment]
@robrap: hello! Thanks for the reminder, sorry I didn't push this forward. A lot of stuff got in the way. Our maintainer team including myself will allocate some time between this week and next to complete this work, at least until its first version.
I'll be tagging you for feedback if you're okay with it! Thanks

**Comment by robrap** [comment]
Hello @mariajgrimaldi. No apologies necessary. I look forward to helping land something.

**Comment by robrap** [comment]
@mariajgrimaldi: Another bit we could add is a reminder to document the event in the service where it is produced, using edx-platform as an example: https://github.com/openedx/edx-platform/blob/f01ef48190c7c4ee6d9e06af4abe21e25fe16dff/docs/guides/hooks/events.rst#index-of-events. 

Related:
1. I never remember where to find the index of events. Maybe just having a note in this doc will be a good enough reminder for me? We'll see.
2. Where is the best place to document that a given event is not only implemented in a service, but is also going to the event bus? Would that just be in an index like the link above? Maybe this depends on if we move this to be part of the event definition? At this point, it is just based on someone adding a listener that sends to the event bus.

File: docs/how-tos/adding-events-to-event-bus.rst
Comment (mariajgrimaldi) [inline docs/how-tos/adding-events-to-event-bus.rst:6]: @robrap @rgraber: I was checking [this confluence](https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus) page to add it here as part of the documentation. But then I wandered, does this page change too much? Should I link that confluence page while it gets more stable? What do you think?

**Comment by mariajgrimaldi** [comment]
I'm finally setting aside some time to do this! Thank you so much for your patience.

**Comment by mariajgrimaldi** [comment]
@robrap:

> Another bit we could add is a reminder to document the event in the service where it is produced, using edx-platform as an example: https://github.com/openedx/edx-platform/blob/f01ef48190c7c4ee6d9e06af4abe21e25fe16dff/docs/guides/hooks/events.rst#index-of-events.

Consider it done!

> Where is the best place to document that a given event is not only implemented in a service but is also going to the event bus? Would that just be in an index like the link above? Maybe this depends on if we move this to be part of the event definition? At this point, it is just based on someone adding a listener that sends to the event bus.

I added a how-to called _Adding a new event to the event bus_, what I think we could do is add an extra step to _How to add an Open edX Event to a service_ saying _Add event to the event to the event bus_ if supported and linking it to the how-to. 

Also, we could modify that index with a column saying "sent over the event bus" that says yes or no. What do you think?

File: docs/how-to/create-new-event.rst
Comment (mariajgrimaldi) [reply docs/how-to/create-new-event.rst]: We currently use the same amount of data when notifying, and for state transfer, I believe the only data class with old / new state is CertificateData, but it's not used. What do you think would be worth mentioning? Should we add optional states to the existing data classes? 🤔

**Comment by mariajgrimaldi** [comment]
@felipemontoya: I would love your input here. Should we add something else?

**Comment by robrap** [comment]
@mariajgrimaldi @rgraber: I'm not certain when I'll get back to review, but I don't want to be a blocker.  I added this to our board, but I'm going to be away for a bit. I recommend we err on landing this quickly, and ticketing up any follow-up that we want to add, but doesn't need to block this initial version. Thanks.

File: docs/how-tos/using-events-within-edx-platform.rst
Comment (mariajgrimaldi) [inline docs/how-tos/using-events-within-edx-platform.rst:1]: note to author: consider generalizing and not use edx-platform-specific instructions. Maybe as an example

File: docs/quickstarts/event-bus.rst
Comment (mariajgrimaldi) [inline docs/quickstarts/event-bus.rst:1]: note to author: take xblock events as example

File: docs/how-tos/creating-new-events.rst
Comment (robrap) [inline docs/how-tos/creating-new-events.rst:24]: Thank you for adding all of these details about subdomains.

I wonder if we could:
1. Keep the table limited to the subdomains in use in this project, and
2. Maybe warn that new subdomains may require some discussion, because there does not yet exist and agreed upon set on subdomains.

Also see this possible adjusted text:
```suggestion
For those reasons, after studying your new event purpose, you must place it in one of the subdomains already in use, or introduce a new subdomain:
```

File: docs/how-tos/creating-new-events.rst
Comment (robrap) [inline docs/how-tos/creating-new-events.rst:29]: 1. Should we call this "Content Authoring" for consistency?
2. Although I know what you mean, I wonder if we should drop "discover" from this description, just to not add confusion with a potential future "discovery" subdomain.

File: docs/how-tos/creating-new-events.rst
Comment (robrap) [inline docs/how-tos/creating-new-events.rst:36]: I think these would be dropped if we simply want to document subdomains in use, and not necessarily those that may be used in the future. That way this doc doesn't have to become the place to debate future subdomains.

File: docs/how-to/create-new-event.rst
Comment (robrap) [reply docs/how-to/create-new-event.rst]: I'm going to mark this resolved. I like what you did, and have new comments on the current iteration.

File: docs/how-to/create-new-event.rst
Comment (robrap) [reply docs/how-to/create-new-event.rst]: I am marking this as resolved. I don't think we did anything with this, but it isn't a blocker for all the great new docs.

File: docs/how-to/create-new-event.rst
Comment (robrap) [reply docs/how-to/create-new-event.rst]: I am marking this as resolved, because I'd love to land these docs sooner rather than later, and I don't think this is a blocker. If anyone wants to research this and iterate on the documentation, that is fine.

File: docs/how-tos/adding-events-to-event-bus.rst
Comment (robrap) [reply docs/how-tos/adding-events-to-event-bus.rst:6]: Linking to this doc would be great. It may change, but I think it is a stable landing page for working with the event bus.

File: docs/how-tos/creating-new-events.rst
Comment (robrap) [inline docs/how-tos/creating-new-events.rst:5]:
```suggestion
Therefore, we've put together a guide on how to add a new event to the library so future contributions are effective.
```

File: docs/how-tos/creating-new-events.rst
Comment (robrap) [inline docs/how-tos/creating-new-events.rst:8]: [thoughts] I'm fine with this landing as-is, and iterating as needed, but @ormsbee was thinking more proprietary events might still live in this repo, possibly with a company specific part of the domain. Don't let this be a blocker to landing this PR. This is just a possible discussion for the future.

File: docs/how-tos/creating-new-events.rst
Comment (robrap) [inline docs/how-tos/creating-new-events.rst:48]:
```suggestion
- Are any of them optional/required?
```

File: docs/how-tos/using-events.rst
Comment (robrap) [inline docs/how-tos/using-events.rst:53]: Should we warn that any non-trivial work should be done in an asynchronous worker task? I think @ormsbee favored any work, because even trivial work may slowly become non-trivial over time. Again, not a blocker, but if we could agree and add sooner than later, that would be nice.

File: docs/quickstarts/event-bus.rst
Comment (robrap) [inline docs/quickstarts/event-bus.rst:4]: Should this doc be deleted, or do you have a plan for it?

File: docs/how-tos/adding-events-to-event-bus.rst
Comment (mariajgrimaldi) [reply docs/how-tos/adding-events-to-event-bus.rst]: good! thank you

**Comment by mariajgrimaldi** [comment]
@robrap: I started addressing your comments! Thank you! I'll let you know when this is ready for review again

File: docs/concepts/hooks-extension-framework.rst
Comment (mariajgrimaldi) [reply docs/concepts/hooks-extension-framework.rst:12]: This should refer to the Open edX platform, I'll change it! Thanks for noticing

Comment (mariajgrimaldi) [reply docs/concepts/hooks-extension-framework.rst:12]: I changed references where it made sense from edx-platform to Open edX platform

File: docs/how-tos/adding-events-to-a-service.rst
Comment (mariajgrimaldi) [reply docs/how-tos/adding-events-to-a-service.rst:4]: I reworded this, so it reads a bit better now. Thanks!

File: docs/how-tos/creating-new-events.rst
Comment (mariajgrimaldi) [reply docs/how-tos/creating-new-events.rst:47]: ups! thanks again!

File: docs/quickstarts/event-bus.rst
Comment (mariajgrimaldi) [reply docs/quickstarts/event-bus.rst]: Deleted!

File: docs/reference/oep-51.rst
Comment (mariajgrimaldi) [reply docs/reference/oep-51.rst:1]: Same here, I added some info!

File: docs/how-tos/creating-new-events.rst
Comment (mariajgrimaldi) [reply docs/how-tos/creating-new-events.rst:8]: What about this for now:

1. Adding optional to the step
2. Adding at the end:
_This step might be unnecessary for events not facing the Open edX community._

File: docs/how-tos/using-events.rst
Comment (mariajgrimaldi) [reply docs/how-tos/using-events.rst]: Yes, I agree that we should encourage the use of that pattern. I added a short warning, what do you think?

File: docs/quickstarts/use-events-to-call-webhook.rst
Comment (mariajgrimaldi) [reply docs/quickstarts/use-events-to-call-webhook.rst:92]: I added a few lines encouraging the use of tasks, what do you think?

**Comment by mariajgrimaldi** [comment]
@robrap: thank you so much! I think I addressed all your comments. Let me know what you think

**Comment by mariajgrimaldi** [comment]
I'll be fixing the failing tests later 😅

File: docs/how-tos/creating-new-events.rst
Comment (robrap) [reply docs/how-tos/creating-new-events.rst:8]: 1. Maybe land this as-is for now?
2. I wonder if we should create an open github issue and point to it in this doc with something like:
```
.. note::
    There is an open discussion about whether organization scoped events would also exist in openedx-events,
    in which case this step would be optional. See <insert issue link here>.
```

**Review by robrap — COMMENTED** [review]
Thanks again @mariajgrimaldi.

File: docs/how-tos/creating-new-events.rst
Comment (mariajgrimaldi) [reply docs/how-tos/creating-new-events.rst:8]: That sounds good for now! 

BTW, there's an open issue about private event schemas that might also be related to this conversation. I'll leave it here just in case: https://github.com/openedx/openedx-events/issues/184

**Comment by mariajgrimaldi** [comment]
@robrap: again, thank you so much for the help! and the patience 😅

File: docs/quickstarts/index.rst
Comment (robrap) [inline docs/quickstarts/index.rst]: @mariajgrimaldi: What happened to the event-bus doc? Did you move it to a new branch, or just kill it?

Comment (robrap) [reply docs/quickstarts/index.rst]: Please ignore. I was looking at a diff, and not the full PR.

**Review by robrap — APPROVED** [review]
Looks great. We can always iterate as-needed. Thanks so much for all the great work!

Not sure if you need to kick off tests again to see if `openedx/cla` check will work?

**Comment by openedx-webhooks** [comment]
@mariajgrimaldi 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-events&entry.867055334=https://github.com/openedx/openedx-events/pull/80&entry.1484655318=https://github.com/mariajgrimaldi&entry.752974735=2022-07-20+17:18&entry.1917517419=2023-04-05+18:08&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
